scripts/compare_benchmarks.py

```diff
@@ -11,141 +14,388 @@
+        value = new_config["context"][key]
+        table_lines.append([colored("≠" + key, "red"), "undefined", value])
+
+    table = AsciiTable(table_lines)
+    table.title = "Configuration Overview"
+    print(table.table)
+
+
+# Doubles the separators (can be | normal rows and + for horizontal separators within the table) given by the list
+# vertical_separators_to_duplicate. [0, 3] means that the first and fourth separator are doubled.
+def double_vertical_separators(lines, vertical_separators_to_duplicate):
+    for line_id, line in enumerate(lines):
+        vertical_separator = line[0]
+        # positions might change due to color coding
+        pos_separators = [
+            m.start() for m in re.finditer(re.escape(vertical_separator), line)
+        ]
+        # 0 required for splicing
+        pos_splits = [0] + [
+            pos_separators[index] for index in vertical_separators_to_duplicate
+        ]
+        new_line = [line[i:j] for i, j in zip(pos_splits, pos_splits[1:] + [None])]
+        lines[line_id] = vertical_separator.join(new_line)
+    return lines
 
 
-if (not len(sys.argv) in [3, 4]):
+if not len(sys.argv) in [3, 4]:
     exit("Usage: " + sys.argv[0] + " benchmark1.json benchmark2.json [--github]")
 
 # Format the output as a diff (prepending - and +) so that Github shows colors
-github_format = bool(len(sys.argv) == 4 and sys.argv[3] == '--github')
+github_format = bool(len(sys.argv) == 4 and sys.argv[3] == "--github")
 
 with open(sys.argv[1]) as old_file:
     old_data = json.load(old_file)
 
 with open(sys.argv[2]) as new_file:
     new_data = json.load(new_file)
 
-if old_data['context']['benchmark_mode'] != new_data['context']['benchmark_mode']:
+if old_data["context"]["benchmark_mode"] != new_data["context"]["benchmark_mode"]:
     exit("Benchmark runs with different modes (ordered/shuffled) are not comparable")
 
-diffs = []
+diffs_throughput = []
+total_runtime_old = 0
+total_runtime_new = 0
 
-table_data = []
-table_data.append(["Benchmark", "prev. iter/s", "runs", "new iter/s", "runs", "change [%]", "p-value"])
+add_note_for_capped_runs = False
+add_note_for_insufficient_pvalue_runs = False
 
-for old, new in zip(old_data['benchmarks'], new_data['benchmarks']):
-    name = old['name']
-    if old['name'] != new['name']:
-        name += ' -> ' + new['name']
+print_context_overview(old_data, new_data)
 
-    if float(old['items_per_second']) > 0.0:
-        diff = float(new['items_per_second']) / float(old['items_per_second'])
-        diffs.append(diff)
+table_data = []
+# $latency and $thrghpt (abbreviated to keep the column at a max width of 8 chars) will later be replaced with a title
+# spanning two columns
+table_data.append(
+    ["Item", "$latency", "", "Change", "$thrghpt", "", "Change", "p-value"]
+)
+table_data.append(["", "old", "new", "", "old", "new", "", ""])
+
+for old, new in zip(old_data["benchmarks"], new_data["benchmarks"]):
```

Comment:
I don't think we should comment on basic Python idioms such as `zip`. I know it might be confusing for people being new to Python , but I couldn't draw a useful line where we should stop explaining basic patterns.

Reply:
Ok, I agree.